rust/kernel/io_mem.rs

```diff
@@ -0,0 +1,195 @@
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the read go out of bounds
+            // (including the type size).
+            unsafe { concat_idents!(rust_helper_, $name)(ptr as _) }
+        }
+
+        /// Reads IO data from the given offset.
+        ///
+        /// It fails if/when the offset (plus the type size) is out of bounds.
+        pub fn $try_name(&self, offset: usize) -> Result<$type_name> {
+            if !Self::offset_ok::<$type_name>(offset) {
+                return Err(Error::EINVAL);
+            }
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // returns an error if `offset` would make the read go out of bounds (including the
+            // type size).
+            Ok(unsafe { concat_idents!(rust_helper_, $name)(ptr as _) })
+        }
+    };
+}
+
+macro_rules! define_write {
+    ($name:ident, $try_name:ident, $type_name:ty) => {
+        /// Writes IO data to the given offset, known at compile time.
+        ///
+        /// If the offset is not known at compile time, the build will fail.
+        pub fn $name(&self, value: $type_name, offset: usize) {
+            Self::check_offset::<$type_name>(offset);
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the write go out of bounds
+            // (including the type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) }
+        }
+
+        /// Writes IO data to the given offset.
+        ///
+        /// It fails if/when the offset (plus the type size) is out of bounds.
+        pub fn $try_name(&self, value: $type_name, offset: usize) -> Result {
+            if !Self::offset_ok::<$type_name>(offset) {
+                return Err(Error::EINVAL);
+            }
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // returns an error if `offset` would make the write go out of bounds (including the
+            // type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) };
+            Ok(())
+        }
+    };
+}
+
+impl<const SIZE: usize> IoMem<SIZE> {
+    /// Tries to create a new instance of a memory block.
+    ///
+    /// The resource described by `res` is mapped into CPU's address space so that it can be
+    /// accessed directly. It is also consumed by this function so that it can't be mapped again
+    /// to a different address.
+    pub fn try_new(res: Resource) -> Result<Self> {
```

Comment:
Shouldn't this be unsafe? The caller needs to guarantee that it "owns" the resource that `Resource` describes.

Reply:
Also, the IO memory mapped cannot trigger any DMA operation, otherwise DMA could be used to safely circumvent all Rust safety measures.

Reply:
It's taking over the the resource. And the resource doesn't have a public constructor.

Reply:
> Also, the IO memory mapped cannot trigger any DMA operation, otherwise DMA could be used to safely circumvent all Rust safety measures.

Of course we want drivers to be able to start DMA operations through this. What would the point of not allowing that?

The question is: how can we restrict this? My thinking was to leave DMA allocation safe, but getting a "handle" (that would be passed to a device for DMA) to the allocation would be unsafe with the restriction that the allocation should be kept alive until the device completed the DMA. (There's no way for us to know when a device is done, so this operation seems intrinsically unsafe.)

The problem with the strategy above is that a driver could write anything to the device, including arbitrary values as DMA handles, and that could of course violate memory safety. I suppose we could make this function unsafe with the requirement that callers need to be careful with DMA, but that's a big hammer and would make this function unnecessarily unsafe for devices like gpio that are not bus-mastering.

Any other ideas?

Reply:
I don't think it's too big an issue to make this function unsafe. A GPIO driver would only need to say: `// SAFETY: This device cannot perform DMA`.

Reply:
Sure, but it's still a big hammer for bus-mastering devices though: "you need to be disciplined while programming operations involving DMA" is not very Rust-like, we'd be basically reverting to C mode and leaving it up to developers to ensure safety. I don't like that.

Reply:
It's up to the author of DMA-capable devices to ensure that all DMA operations are correct, and I don't think it could be checked. The best we can do is a separate `IoMem` that is safe to create but unsafe to access.

Reply:
I suppose we could make the safety requirement be that either (a) no DMA operations are performed, or (b) DMA buffers should be retrieved via the `dma` module (which will have further requirements).

Reply:
Well, regardless how DMA buffers are retrieved, it's still up to the device to ensure that correct address is written isn't it?

Reply:
It's up to the driver, yes.

But if we make the DMA "get handle" interface unsafe and require developers to go through it, we force them to consider the safety requirements then (when programming an operation). My feeling is that having the safety requirement only in `IoMem::try_new` that is buried in init code rarely looked at will allow developers to violate them without noticing it -- not too different from C.

Reply:
Maybe use macros to generate register maps would be the best approach for safety; some registers could be marked unsafe with this approach. E.g.
```rust
iomap! {
	PL061IoMem {
		  gpiodir: u8 @ 0x400,
		  gpiois: u8 @ 0x404,
		  gpiobe: u8 @ 0x408,
		  gpioiev: u8 @ 0x40C,
		  gpioie: u8 @ 0x410,
		  gpiomis: u8 @ 0x418,
		  gpioic: u8 @ 0x41C,
	}
}
```
and then for a DMA device a register could be written as
```rust
		unsafe dma_addr: u64 @ ...,
```

Reply:
I like where this is going. It also goes well with Miguel's idea of automatically generating accessors for individual registers.

One problem with this approach is that we don't always encode the dma address via IO mem like this. For example, in the NVMe case, we use host memory to hold command queues -- the entries in these queues hold DMA addresses, and IO mem is only a doorbell to indicate that new entries have been added to queues.

Reply:
I have no experience with NVMe but from your description it is does not sound very different from how virtio queue works. The command queue setup registers could be made unsafe in that case, and it's then up to the command queue abstraction to propagate the safety requirements.